python/tskit/genotypes.py

```diff
@@ -224,6 +233,55 @@ def copy(self) -> Variant:
         variant_copy._ll_variant = self._ll_variant.restricted_copy()
         return variant_copy
 
+    def counts(self) -> dict[str, int]:
+        """
+        Return a dictionary mapping each possible :attr:`allele <Variant.alleles>`
+        at this site to the number of samples possessing that allele among the
+        set of samples specified when creating this Variant (by default, this is all
+        the sample nodes in the tree sequence). Missing data is represented by an
+        allelic state of ``None``.
+
+        :return: A dictionary mapping each allelic state to the number of samples
+            with that allele.
+        """
+        counts = {a: 0 for a in self.alleles}
+        for g, count in zip(*np.unique(self.genotypes, return_counts=True)):
```

Comment:
This seems like an inefficient way do this to me. unique requires a full sort, O(n log n), but we can also do this with num_alleles passes through the array, like
```
for j, allele in self.alleles:
      counts[allele] = np.sum(self.genotypes == j)
```

since num_alleles is typically small, this should be faster for large n.

Reply:
Good point. Happy to go with this. Or we could use `np.bincount` since these are all likely to be integers from 0..n:

```python
bincounts = np.bincount
for j, allele in enumerate(self.alleles):
    counts[allele] = bin counts[j]
```

?

Reply:
Will bincount work with ``-1``?

Reply:
Ah, really good point! I forgot about -1.

Reply:
We could hack it by adding 1, but then I think your suggestion is much cleaner.

Reply:
I had to special-case the `-1` anyway, because you can't simply loop through the alleles and treat the indices as required: if the last element is `None` you need to check for `-1` in the array.

So I used bin count (which is cleaner and I guess a bit faster) when there is no missing data, and a modification of your code when there is missing data. It's not that clean, but the clearest I can come up with.